Ladybird/Android/src/main/cpp/WebViewImplementationNative.cpp

```diff
@@ -4,31 +4,78 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <LibGfx/Bitmap.h>
+#include <LibGfx/Painter.h>
 #include <LibWebView/ViewImplementation.h>
+#include <android/bitmap.h>
 #include <android/log.h>
 #include <jni.h>
 
-using namespace WebView;
-
 namespace {
+
+Gfx::BitmapFormat to_gfx_bitmap_format(i32 f)
+{
+    switch (f) {
+    case ANDROID_BITMAP_FORMAT_RGBA_8888:
+        return Gfx::BitmapFormat::BGRA8888;
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
 class WebViewImplementationNative : public WebView::ViewImplementation {
 public:
-    virtual Gfx::IntRect viewport_rect() const override { return {}; }
-    virtual Gfx::IntPoint to_content_position(Gfx::IntPoint) const override { return {}; }
-    virtual Gfx::IntPoint to_widget_position(Gfx::IntPoint) const override { return {}; }
+    virtual Gfx::IntRect viewport_rect() const override { return m_viewport_rect; }
+    virtual Gfx::IntPoint to_content_position(Gfx::IntPoint p) const override { return p; }
+    virtual Gfx::IntPoint to_widget_position(Gfx::IntPoint p) const override { return p; }
     virtual void update_zoom() override { }
 
+    void paint_into_bitmap(void* android_bitmap_raw, AndroidBitmapInfo const& info)
+    {
+        // Software bitmaps only for now!
+        VERIFY((info.flags & ANDROID_BITMAP_FLAGS_IS_HARDWARE) == 0);
+
+        auto android_bitmap = MUST(Gfx::Bitmap::create_wrapper(to_gfx_bitmap_format(info.format), { info.width, info.height }, 1, info.stride, android_bitmap_raw));
+        Gfx::Painter painter(android_bitmap);
+        if (auto* bitmap = m_client_state.has_usable_bitmap ? m_client_state.front_bitmap.bitmap.ptr() : m_backup_bitmap.ptr())
+            painter.blit({ 0, 0 }, *bitmap, bitmap->rect());
+        else
+            painter.clear_rect(painter.clip_rect(), Gfx::Color::Magenta);
+
+        // Convert our internal BGRA into RGBA. This will be slowwwwwww
+        // FIXME: Don't do a color format swap here.
+        for (auto y = 0; y < android_bitmap->height(); ++y) {
+            auto* scanline = android_bitmap->scanline(y);
+            for (auto x = 0; x < android_bitmap->width(); ++x) {
+                auto current_pixel = scanline[x];
+                u32 alpha = (current_pixel & 0xFF000000U) >> 24;
+                u32 red = (current_pixel & 0x00FF0000U) >> 16;
+                u32 green = (current_pixel & 0x0000FF00U) >> 8;
+                u32 blue = (current_pixel & 0x000000FFU);
+                scanline[x] = (alpha << 24U) | (blue << 16U) | (green << 8U) | red;
+            }
+        }
+    }
+
+    void set_viewport_geometry(int w, int h)
+    {
+        m_viewport_rect = { { 0, 0 }, { w, h } };
```

Comment:
That doesn't look right. You should pass the whole rect from the Android view.

Reply:
No, this should account for scrolling too. So for example if you have scrolled 200 px towards bottom, 30px right, and your `WebView` size is 500x400, the rect you should pass here will be `{ { 200, 30 }, { 500, 400 } }`

Here's the corresponding place In Ladybird GTK 4: https://github.com/bugaevc/ladybird-gtk4/blob/fadc9947535ec1d2f84e4ca52f9d5d70e25cf0b5/WebView.cpp#L499C46-L499C46

Reply:
I've added a FIXME for this to WebView.kt, since the view doesn't support scrolling yet